app/src/main/java/id_iot/demo/CollectionAdapter.java

```diff
@@ -0,0 +1,107 @@
+package id_iot.demo;
+
+import android.content.Context;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.BaseAdapter;
+import android.widget.Filterable;
+import android.widget.ImageView;
+import android.widget.TextView;
+import android.widget.Filter;
+import java.util.ArrayList;
+import java.util.List;
+
+public class CollectionAdapter extends BaseAdapter implements Filterable {
+    
+    private Context context;
+    private ArrayList<Recipe> recipes;
+    private ArrayList<Recipe> backupRecipes;
+    private ValueFilter valueFilter;
+
+    public CollectionAdapter(Context context, ArrayList<Recipe> recipes) {
+        this.context = context;
+        this.recipes = recipes;
+        this.backupRecipes = recipes;
+    }
+
+    @Override
+    public long getItemId(int position) {
+        return position;
+    }
+
+    @Override
+    public Object getItem(int position) {
+        return recipes.get(position);
+    }
+
+    @Override
+    public int getCount() {
+        return recipes.size();
+    }
+
+    public View getView(int position, View convertView, ViewGroup parent) {
+        final LayoutInflater layoutInflater = LayoutInflater.from(context);
+        convertView = layoutInflater.inflate(R.layout.collection_view_cell, null);
+
+        ImageView recipePhoto = (ImageView) convertView.findViewById(R.id.collection_image);
+        ImageView favoriteStar = (ImageView) convertView.findViewById(R.id.collection_favorite);
+        TextView recipeName = (TextView) convertView.findViewById(R.id.collection_name);
+        TextView recipeDetail = (TextView) convertView.findViewById(R.id.collection_detail);
+        recipeName.setText(recipes.get(position).getRecipeName());
+        recipeDetail.setText(recipes.get(position).getRecipeDetail());
+        favoriteStar.setImageResource(recipes.get(position).getFavorite()? R.drawable.favorite_checked: R.drawable.favorite_unchecked);
+        recipePhoto.setImageResource(recipes.get(position).getRecipePhoto());
+        return convertView;
+    }
+
+    public ArrayList<Recipe> getRecipes(){
+        return this.recipes;
+    }
+
+    @Override
+    public Filter getFilter() {
+        if (valueFilter == null) {
+            valueFilter = new ValueFilter();
+        }
+        return valueFilter;
+    }
+
+    private class ValueFilter extends Filter {
+
+        @Override
+        protected FilterResults performFiltering(CharSequence constraint) {
+            FilterResults results = new FilterResults();
+
+            if (!constraint.toString().isEmpty()){
```

Comment:
actually usually first goes positive case. I mean if there is else statement don't start from ! condition. But up to you. It is just later really easy to miss as you expect to see positive first